<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Output file format &mdash; GiRAFR 1.0.0 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/sphinx_highlight.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="Detect CRISPR Cas9 editing effect" href="editing_effect.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="index.html" class="icon icon-home"> GiRAFR
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="gRNA_mutation.html">Single cell CRISPR screen gRNA library mutation profiling</a></li>
<li class="toctree-l1"><a class="reference internal" href="editing_effect.html">Detect CRISPR Cas9 editing effect</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Output file format</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#grna-mutation">gRNA_mutation</a></li>
<li class="toctree-l2"><a class="reference internal" href="#editing-effect">editing_effect</a></li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">GiRAFR</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
      <li>Output file format</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/output.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <div class="section" id="output-file-format">
<h1>Output file format<a class="headerlink" href="#output-file-format" title="Permalink to this heading"></a></h1>
<div class="section" id="grna-mutation">
<span id="output"></span><h2>gRNA_mutation<a class="headerlink" href="#grna-mutation" title="Permalink to this heading"></a></h2>
<p>consensus.bam:
Bam file contains consensus sequence alignment result for each UMI.</p>
<p>consensus.sequence.gRNA.variant.txt</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 50%" />
<col style="width: 50%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>column number</p></th>
<th class="head"><p>Value</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Column 1</p></td>
<td><p>Cell barcode</p></td>
</tr>
<tr class="row-odd"><td><p>Column 2</p></td>
<td><p>UMI</p></td>
</tr>
<tr class="row-even"><td><p>Column 3</p></td>
<td><p>Number of UMI in the cell</p></td>
</tr>
<tr class="row-odd"><td><p>Column 4</p></td>
<td><p>Number of reads for this UMI (default min = 1)</p></td>
</tr>
<tr class="row-even"><td><p>Column 5</p></td>
<td><p>single/multiple. Single refers to UMI with only one type of sequence. Multiple refers to UMI with more than one type of sequence.</p></td>
</tr>
<tr class="row-odd"><td><p>Column 6</p></td>
<td><p>Consensus sequence. Sequence with most reads supported.</p></td>
</tr>
<tr class="row-even"><td><p>Column 7</p></td>
<td><p>Number of reads that support the consensus sequence.</p></td>
</tr>
<tr class="row-odd"><td><p>Column 8</p></td>
<td><p>True/False. True: wild type sequence. False: different from reference.</p></td>
</tr>
<tr class="row-even"><td><p>Column 9</p></td>
<td><p>Name of the guide. ‘multiple gene’: when UMI mapped to multiple guides with equal number of reads. The progam cannot assign gene to UMI confidently. In this case, UMI will be skipped later.</p></td>
</tr>
<tr class="row-odd"><td><p>Column 10</p></td>
<td><p>Guide variant type with detailed mismatch information</p></td>
</tr>
<tr class="row-even"><td><p>Column 11</p></td>
<td><p>Mutation structure annotation</p></td>
</tr>
</tbody>
</table>
<p>cells.gRNA.txt</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 50%" />
<col style="width: 50%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>column number</p></th>
<th class="head"><p>Value</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Column 1</p></td>
<td><p>Cell barcode</p></td>
</tr>
<tr class="row-odd"><td><p>Column 2</p></td>
<td><p>Number of guides in the cell</p></td>
</tr>
<tr class="row-even"><td><p>Column 3</p></td>
<td><p>Guide name.</p></td>
</tr>
<tr class="row-odd"><td><p>Column 4</p></td>
<td><p>WT: Intact gRNA</p></td>
</tr>
<tr class="row-even"><td><p>Column 5</p></td>
<td><p>Number of UMI support wild type guide</p></td>
</tr>
<tr class="row-odd"><td><p>Column 6</p></td>
<td><p>Guide variant type if exist.</p></td>
</tr>
<tr class="row-even"><td><p>Column 7</p></td>
<td><p>Number of UMI support mutant guide.</p></td>
</tr>
</tbody>
</table>
<p>cells.gRNA.single.txt: cells with only one gRNA</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 50%" />
<col style="width: 50%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>column number</p></th>
<th class="head"><p>Value</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Column 1</p></td>
<td><p>Cell barcode</p></td>
</tr>
<tr class="row-odd"><td><p>Column 2</p></td>
<td><p>1 as only one guide in the cell</p></td>
</tr>
<tr class="row-even"><td><p>Column 3</p></td>
<td><p>Guide name. Guide name ends with WT represents wild type.</p></td>
</tr>
<tr class="row-odd"><td><p>Column 4</p></td>
<td><p>Number of UMI support mutant guide.</p></td>
</tr>
</tbody>
</table>
<p>cells.gRNA.single.MT.txt</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 50%" />
<col style="width: 50%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>column number</p></th>
<th class="head"><p>Value</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Column 1 to Column 4</p></td>
<td><p>same as cells.gRNA.single.txt</p></td>
</tr>
<tr class="row-odd"><td><p>Column 5</p></td>
<td><p>Mutation details. None when wildtype or soft clipping only.</p></td>
</tr>
<tr class="row-even"><td><p>Column 6</p></td>
<td><p>Mutant sequence. None when wildtype or soft clipping only.</p></td>
</tr>
</tbody>
</table>
<p>all.MT.txt: mutations in single cell in a long format, each line contains single mutation event.</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 50%" />
<col style="width: 50%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>column number</p></th>
<th class="head"><p>Value</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Column 1</p></td>
<td><p>target gene</p></td>
</tr>
<tr class="row-odd"><td><p>Column 2</p></td>
<td><p>structure (defined in structure.gtf)</p></td>
</tr>
<tr class="row-even"><td><p>Column 3</p></td>
<td><p>length of the structure</p></td>
</tr>
<tr class="row-odd"><td><p>Column 4</p></td>
<td><p>Position on the structure</p></td>
</tr>
<tr class="row-even"><td><p>Column 5</p></td>
<td><p>mutation (<a class="reference internal" href="gRNA_mutation.html#cigar-string"><span class="std std-ref">CIGAR similar format</span></a>)</p></td>
</tr>
<tr class="row-odd"><td><p>Column 6</p></td>
<td><p>variant name with number suffix.</p></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="editing-effect">
<h2>editing_effect<a class="headerlink" href="#editing-effect" title="Permalink to this heading"></a></h2>
<p>For each gRNA:
<code class="docutils literal notranslate"><span class="pre">{region_name}.bam</span></code>: alignment within detection window
<code class="docutils literal notranslate"><span class="pre">{region_name}.bam.featureCounts.sorted.bam</span></code>: FeatureCount assigned reads to gene.
<code class="docutils literal notranslate"><span class="pre">{region_name}.consensus.sequence.txt</span></code>: Consensus sequence within detection window for each UMI and cell.</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 50%" />
<col style="width: 50%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>column number</p></th>
<th class="head"><p>Value</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Column 1</p></td>
<td><p>cell barcode</p></td>
</tr>
<tr class="row-odd"><td><p>Column 2</p></td>
<td><p>UMI</p></td>
</tr>
<tr class="row-even"><td><p>Column 3</p></td>
<td><p>Number of UMI for cell</p></td>
</tr>
<tr class="row-odd"><td><p>Column 4</p></td>
<td><p>Number of sequences for that UMI</p></td>
</tr>
<tr class="row-even"><td><p>Column 5</p></td>
<td><p>reference sequence in the detection window</p></td>
</tr>
<tr class="row-odd"><td><p>Column 6</p></td>
<td><p>union sequence for that UMI in the detection window</p></td>
</tr>
<tr class="row-even"><td><p>Column 7</p></td>
<td><p>union number of reads</p></td>
</tr>
<tr class="row-odd"><td><p>Column 8</p></td>
<td><p>Boolean True for wildtype sequence, false for mutation exist in sequence</p></td>
</tr>
<tr class="row-even"><td><p>Column 9</p></td>
<td><p>mapped gene</p></td>
</tr>
<tr class="row-odd"><td><p>Column 10</p></td>
<td><p>mapped positions on the genome</p></td>
</tr>
<tr class="row-even"><td><p>Column 11</p></td>
<td><p>is_umi_seq_different</p></td>
</tr>
<tr class="row-odd"><td><p>Column 12</p></td>
<td><p>reference gRNA id (region_name)</p></td>
</tr>
</tbody>
</table>
<p>{region_name}.editing_effect.cutsite.txt</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 50%" />
<col style="width: 50%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>column number</p></th>
<th class="head"><p>Value</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Column 1 to Column 12</p></td>
<td><p>same as {region_name}.consensus.sequence.txt</p></td>
</tr>
<tr class="row-odd"><td><p>Column 13</p></td>
<td><p>mutation, ‘No mutation’ given when wildtype sequence. “Deletion at cutsite”, “Insertion/Deletion somewhere else”, “Mismatch”</p></td>
</tr>
<tr class="row-even"><td><p>Column 14</p></td>
<td><p>gRNA assigned to this cell</p></td>
</tr>
</tbody>
</table>
</div>
</div>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="editing_effect.html" class="btn btn-neutral float-left" title="Detect CRISPR Cas9 editing effect" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Qian Yu.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>